Use target attitude consistently in roll optimization

taldcroft opened the pull request:
This makes the changes I proposed in #95.  The code changes are actually less than they look because I took the opportunity to make some variable naming more consistent.

But the very important one is that "roll" in the roll info output is **always** the target attitude roll.   For ERs that is the same as the ACA attitude. That's basically what the mission planners want to see.

So by design now there will be a diff between "roll" in the roll options and `acar.roll`, but the roll option roll will match `acar.att_targ.roll`.  Likewise `roll_min` and `roll_max` refer to target attitude roll.

Fixes #95 

cc: @jskrist

taldcroft commented:
This passes unit tests but I haven't tried digging in.  I need to step away for a bit, but maybe @jeanconn will have a look and try to break it with an 89.9 deg declination case?  :smile:

jeanconn commented on sparkles/core.py:
```diff
@@ -456,6 +456,12 @@ def review_status(self):
+    @property
+    def att_targ(self):
+        if not hasattr(self, '_att_targ'):
+            self._att_targ = self._calc_targ_from_aca(self.att, 0, 0)
```
This looks good (computing the _att_targ as needed) though I suppose if we get the y_off z_off values there will be more places to scrub.  Also, not sure why _calc_targ_from_aca requires args at all, but that's not an issue for this PR.

taldcroft replied:
Agreed those should be optional kwargs.  When we get offsets in the code here it will be an easy scrub.

jeanconn commented:
I can't make it error out with an 89.99 dec (or -89.99) case, but not sure what to add to the tests.

taldcroft commented:
@jeanconn - I'm working on a functional test now, so you don't need to do any more on this for now (in terms of testing).

taldcroft commented:
See:

https://icxc.cfa.harvard.edu/aspect/tmp/obs18000/
https://icxc.cfa.harvard.edu/aspect/tmp/obs48000/

These are the outputs for an ER and OR at a dec=89.9, where the difference between ACA and target frames is significant.  If you want to play along at home:
```
from proseco import get_aca_catalog
from proseco.tests.test_common import mod_std_info
from Ska.Sun import nominal_roll
from sparkles import run_aca_review
from Quaternion import Quat

dec = 89.9
date = '2019:006'
roll = nominal_roll(0, dec, time=date)
att = Quat([0, dec, roll])
print(f'nominal_roll = {roll} == {att.roll}')

obsid = 18000
label = f'obs{obsid}'

kwargs = mod_std_info(att=att, n_guide=8, obsid=obsid, date=date)
aca = get_aca_catalog(**kwargs)

acar = aca.get_review_table()

run_aca_review(acars=[acar], make_html=True, load_name=label,
               report_dir=label, roll_level='all')
```

taldcroft commented:
And yes, as predicted, I found some oddity in this high-dec case where it finds a roll option if it is labeled as an ER but not when labeled as OR.  Will investigate!

jeanconn commented:
It seems to me that is possible (when rotating about slightly different
center).

On Thu, Feb 28, 2019, 3:28 PM Tom Aldcroft <notifications@github.com> wrote:

> And yes, as predicted, I found some oddity in this high-dec case where it
> finds a roll option if it is labeled as an ER but not when labeled as OR.
> Will investigate!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sot/sparkles/pull/97#issuecomment-468425944>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwT1KCH3Z_cfDJ7yht1n-7yLZ1bMWy2ks5vSDwDgaJpZM4bXBK->
> .
>

taldcroft commented:
Actually my code is fine, the issue was that it uses `n_stars` as part of the metric to decide if a new roll is better.  However `n_stars` here is `guide_count` for ORs (roughly the count of 10th mag or brighter stars) and it is `guide_count_9th` for ERs (roughly the count of 9th mag or brighter stars).  

I think these counts are the right ones for evaluating ORs and ERs respectively, but it can be confusing.

So probably just add a test and we're good?

jeanconn commented on sparkles/roll_optimize.py:
```diff
@@ -150,39 +150,46 @@ def get_roll_intervals(self, cand_idxs, roll_nom=None, roll_dev=None,
         if roll_dev is None:
             roll_dev = allowed_rolldev(pitch)
 
         # Ensure roll_nom in range 0 <= roll_nom < 360 to match q_att.roll.
```
Nit: comment references q_att.roll .

jeanconn commented on sparkles/roll_optimize.py:
```diff
@@ -150,39 +150,46 @@ def get_roll_intervals(self, cand_idxs, roll_nom=None, roll_dev=None,
+            roll_nom = Ska.Sun.nominal_roll(att.ra, att.dec, self.date)
+            att_nom = Quat([att.ra, att.dec, roll_nom])
+            att_nom_targ = self._calc_targ_from_aca(att_nom, y_off, z_off)
+            roll_nom = att_nom_targ.roll
```
Another nit:  docstring should probably mention that roll_nom should be supplied in target frame (not that it is really ever supplied).

jeanconn commented:
I did another pass through the code and it makes sense to me and I think it is good to go.  Addressing nits optional. 

The logical intervals handling is a little fancy for me, but the rolls look consistently to be supplied in target frame and then transformed to ACA frame to actually check catalogs (which makes sense to me).

taldcroft commented:
OK, I've update the script and report HTML pages from https://github.com/sot/sparkles/pull/97#issuecomment-468417949.  Please re-review and see if the numbers all seem to make sense.

I excluded a bunch of stars from the initial catalog to make it worse and thus make the roll options more attractive.

jeanconn commented on sparkles/core.py:
```diff
@@ -622,15 +623,20 @@ def get_text_pre(self):
         catalog = '\n'.join(self.pformat(max_width=-1, max_lines=-1))
         self.acq_count = np.sum(self.acqs['p_acq'])
 
+        att_string = f'ACA RA, Dec, Roll (deg): {att.ra:.6f} {att.dec:.5f} {att.roll:.5f}'
```
Though I suppose I meant to ask if RA really deserved an extra decimal place compared Dec and Roll.

jeanconn commented:
I am not that surprised that test_get_roll_intervals is now failing.

jeanconn commented:
I cut the no-longer-true assert statement and pushed to this branch directly.

jeanconn commented:
I think maybe the roll "mock obsids" in the roll report are still the acar.att.roll values?  A quick search didn't show me where those actually come from so haven't finished with that.

taldcroft commented:
@jeanconn @jskrist - I've done a bunch more clean-up and I think addressed all issues raised so far.  I made the report page more clear and added the roll options summary table to the roll options review page.  See the sample report pages linked above as they have been updated.

Except didn't do the 6 to 5 decimals places.  I agree that 5 is fine, I just didn't get to it.
